README.md
Did Imputation
================

<!-- README.md is generated from README.Rmd. Please edit that file -->

<img src="man/figures/logo.png" width="150px" align="right" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/cdfInnovLab/didimputation/workflows/R-CMD-check/badge.svg)](https://github.com/CdfInnovLab/didimputation/actions)

[![Codecov test
coverage](https://codecov.io/gh/CdfInnovLab/didImputation/branch/master/graph/badge.svg)](https://codecov.io/gh/CdfInnovLab/didImputation?branch=master)

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)

[![](https://img.shields.io/badge/devel%20version-0.0.2-blue.svg)](https://github.com/CdfInnovLab/didImputation)

[![](https://img.shields.io/github/languages/code-size/CdfInnovLab/didImputation.svg)](https://github.com/CdfInnovLab/didImputation)

<!-- badges: end -->

Estimation of staggered Difference-in-Differences using the imputation
approach of Borusyak, Jaravel, and Spiess (2021). The packages allows
for:

-   Multiple time periods

-   Staggered design (i.e., units are treated at different time periods)

-   Continuous controls

The package implements an imputation method to estimate the treatment
effect and pre-trend testing in difference-in-differences designs with
staggered adoption (i.e where units are treated at different time
periods). Recent literature stress out the importance of not using the
standard *twoway fixed effect* regression.

The standard DiD setup involves two periods and two groups (one treated
and one untreated), it relies on parallel trend assumption to estimate
the treatment effect of the treated. The staggered DiD setup is the
generalization of this approach to multiple periods and multiple groups
(i.e. individuals treated at different time periods.). Recent literature
stress out the need to not use the standard two-way fixed effect (TWFE)
regression to estimate those models. This package implements a method of
imputation to estimate the average treatment effect. The package uses
untreated observations to predict the counterfactual outcome on treated
observations and provide the appropriate Standard errors. It provides
ways to test for parallel trends.

## Installation

You can install the github version with `devtools` or
`renv`(recommended, [read more about
renv](https://rstudio.github.io/renv/articles/renv.html))

``` r
devtools::install("CdfInnovLab/didImputation")
```

## Usage

``` r
library(didImputation)
```

``` r
data(did_simulated)

res <- didImputation(y0 = y ~ 0 | i + t,
                    cohort = 'g',
                    data = did_simulated)
summary(res)
#> Event Study: imputation method. Dep. Var.:  y 
#> Counterfactual model:  y ~ 0 | i + t 
#> Number of cohorts:  5 
#> Observations: 1500 
#> |-Treated:    630 
#> |-Untreated:  870 
#>       Estimate Std. Error t value Pr(>|t|))
#> k::-4 0.058    0.228       0.256  0.798    
#> k::-3 0.124    0.238       0.518  0.605    
#> k::-2 0.204    0.257       0.795  0.428    
#> k::-1 0.059    0.284       0.208  0.835    
#> k::0  0.974    0.097      10.016  <0.001***
#> k::1  2.086    0.110      18.972  <0.001***
#> k::2  2.991    0.143      20.899  <0.001***
#> k::3  3.981    0.189      21.046  <0.001***
#> k::4  4.775    0.259      18.463  <0.001***
#> ---
#> Signif. Code: 0 '***' 0.01 '**' 0.05 '*' 0.1 '' 1 
#> Wald stats for pre-trends:
#> Wald (joint nullity): stat = 0.473843, p = 0.754974, on 4 and 860 DoF, VCOV: Clustered (i).
```

You can print the result easily with `didplot`

``` r
didplot(res)
```

<img src="man/figures/README-plot-1.png" width="100%" />

# How it works

didImputation estimates the effects of a binary treatment with staggered
timing. It allows for arbitrary heterogeneity of treatment and dynamic
effects.

The estimation is a three step procedures

1.  **Estimate** a linear model on non treated observations only (it
    ![\\Omega_0](https://latex.codecogs.com/png.latex?%5COmega_0 "\Omega_0"))
    (either not-yet-treated or never-treated).

    ![Y\_{it}(0\|it \\in \\Omega_0) = \\alpha_i + \\beta_t + X\_{it}'\\delta + \\varepsilon\_{it}](https://latex.codecogs.com/png.latex?Y_%7Bit%7D%280%7Cit%20%5Cin%20%5COmega_0%29%20%3D%20%5Calpha_i%20%2B%20%5Cbeta_t%20%2B%20X_%7Bit%7D%27%5Cdelta%20%2B%20%5Cvarepsilon_%7Bit%7D "Y_{it}(0|it \in \Omega_0) = \alpha_i + \beta_t + X_{it}'\delta + \varepsilon_{it}")

2.  **Impute** the treated observations
    (![it \\in \\Omega_1](https://latex.codecogs.com/png.latex?it%20%5Cin%20%5COmega_1 "it \in \Omega_1"))
    potential outcome
    ![Y\_{it}(0)](https://latex.codecogs.com/png.latex?Y_%7Bit%7D%280%29 "Y_{it}(0)")
    and obtain treatment effect
    ![\\tau\_{it}](https://latex.codecogs.com/png.latex?%5Ctau_%7Bit%7D "\tau_{it}")
    by substracting the predicted outcome from step 1

    ![\\begin{align\*}
    \\hat{Y}\_{it}(0\|it \\in \\Omega_1) &= \\hat{\\alpha}\_i + \\hat{\\beta}\_t + X\_{it}'\\hat{\\delta} \\\\
    \\hat{\\tau}\_{it} &= Y\_{it} - \\hat{Y}\_{it}(0)
    \\end{align\*}](https://latex.codecogs.com/png.latex?%5Cbegin%7Balign%2A%7D%0A%5Chat%7BY%7D_%7Bit%7D%280%7Cit%20%5Cin%20%5COmega_1%29%20%26%3D%20%5Chat%7B%5Calpha%7D_i%20%2B%20%5Chat%7B%5Cbeta%7D_t%20%2B%20X_%7Bit%7D%27%5Chat%7B%5Cdelta%7D%20%5C%5C%0A%5Chat%7B%5Ctau%7D_%7Bit%7D%20%26%3D%20Y_%7Bit%7D%20-%20%5Chat%7BY%7D_%7Bit%7D%280%29%0A%5Cend%7Balign%2A%7D "\begin{align*}
    \hat{Y}_{it}(0|it \in \Omega_1) &= \hat{\alpha}_i + \hat{\beta}_t + X_{it}'\hat{\delta} \\
    \hat{\tau}_{it} &= Y_{it} - \hat{Y}_{it}(0)
    \end{align*}")

3.  **Average** estimated treatment effects
    ![\\tau\_{it}](https://latex.codecogs.com/png.latex?%5Ctau_%7Bit%7D "\tau_{it}")
    to the estimand of interest.

    > For the overall average treatment effect, the estimate is defined
    > by
    >
    > ![\\hat{\\tau} = \\sum\_{it \\in \\Omega_1} \\tau\_{it}](https://latex.codecogs.com/png.latex?%5Chat%7B%5Ctau%7D%20%3D%20%5Csum_%7Bit%20%5Cin%20%5COmega_1%7D%20%5Ctau_%7Bit%7D "\hat{\tau} = \sum_{it \in \Omega_1} \tau_{it}")

# TODO

-   [x] Estimation weights
-   [x] Triple differences
-   [x] Vignette
-   [ ] Time invariant controls
-   [ ] Unit invariant controls
-   [ ] Custom cluster
-   [ ] Latex export
-   [ ] Allow custom period length
-   [ ] Automatic panel balance
-   [ ] Interactions in fixed effects
-   [ ] Allow weights reuse

## Reference

[Borusyak, K., Jaravel, X., & Spiess, J. (2021). Revisiting event study
designs: Robust and efficient estimation. Working
paper.](https://www.google.com/url?q=https%3A%2F%2Fwww.dropbox.com%2Fs%2F0o79nppmve792nf%2Fborusyak_hull_jan21.pdf%3Fraw%3D1&sa=D&sntz=1&usg=AFQjCNE2vdSXDowNFgVeRfpaUacGMQop-A)

## See also

[didimputation](https://github.com/kylebutts/didimputation): Another
implementation using sparse matrix inversion.


NEWS.md
# didImputation 0.0.3 (2021-10-09)

## Features

* `didImputation` now supports triple difference estimation with `td` option.

## Bugs

* Rename internal variables to avoid conflict with user input.

## Misc

* Updated `README`.
* A vignette is now available.

# didImputation 0.0.2 (2021-08-05)

## Features

* New function `generateDidData()` to create fake data.
* `didImputation` now supports sample weights.

## Bugs

* Now raise error when `fixest::predict()` cannot extract fixed-effects.

## Misc

* Added a `NEWS.md` file to track changes to the package.
* Update `README` with estimation method.


R/methods.R
#' coef method for object of class "didImputation"
#'
#' @inheritParams stats::coef
#'
#' @export coef.didImputation
#' @export
coef.didImputation <- function(object, ...) {
  return(object$coefs)
}

#' DiD imputation summary result
#'
#' @inheritParams base::summary
#'
#' @importFrom stats symnum
#'
#' @export summary.didImputation
#' @export
summary.didImputation <- function(object, ...) {
  s <- object

  ans <- list(
    dep = s$y,
    call = s$y0,
    Ncohorts = length(unique(s$data[[s$cohort]]))-1,
    residuals = sqrt(s$data$.wei) * s$data$.eps,
    N = nrow(s$data),
    N0 = nrow(s$data[.d == 0]),
    N1 = nrow(s$data[.d == 1]),
    weights = s$data$wei,
    rss = sum(s$data$.wei * s$data$.eps^2),
    horizons = s$coef,
    coeftable = s$coeftable,
    weights_cols = s$data[, .SD, .SDcols = s$weights_cols],
    signif = symnum(s$coeftable$`Pr(>|t|))`, corr = FALSE, na = FALSE,
                     cutpoints = c(0, 0.01, 0.05, 0.1, 1),
                     symbols = c("***", "**", "*", ""))
  )

  if (s$coef[1] < 0) {
    ans$pre_trend <- s$pre_trends
    ans$pre_wald <- s$wald
  }

  class(ans) <- "summary.didImputation"

  return(ans)
}

#' Print method for summary of class "didImputation"
#'
#' @param x A list of class "summary.didImputation"
#' @param digits Decimal precision in output.
#' @param ... additional arguments.
#'
#' @return An estimation table
#' @export print.summary.didImputation
#' @export
print.summary.didImputation <- function(x, digits = 3, ...){
  ans <- x
  dt <- round(ans$coeftable, digits)

  p_threshold <- 1*10^(-digits)
  p_threshold_str <- formatC(p_threshold, digits)

  dt$`Pr(>|t|))` <- ifelse(dt$`Pr(>|t|))` < p_threshold,
                           paste0('<',p_threshold_str),
                           dt$`Pr(>|t|))`)

  dt$`Pr(>|t|))` <- paste0(dt$`Pr(>|t|))`, ans$signif)


  cat("Event Study: imputation method. Dep. Var.: ", ans$dep, "\n")
  cat("Counterfactual model: ", deparse(ans$call), "\n")
  cat("Number of cohorts: ", ans$Ncohorts, "\n")
  cat("Observations:", ans$N, "\n")
  cat("|-Treated:   ", ans$N1, "\n")
  cat("|-Untreated: ", ans$N0, "\n")
  print(dt, right = FALSE)
  cat("---\n")
  cat("Signif. Code:", attr(ans$signif, "legend"), "\n")

  if (any(ans$horizons < 0)) {
    cat("Wald stats for pre-trends:\n")
    print(ans$pre_wald)
  }
}

#' Print didImputation result
#'
#' @param x an object of class "didImputation".
#' @param ... additional arguments.
#'
#' @importFrom utils tail
#'
#' @export print.didImputation
#' @export
print.didImputation <- function(x, ...) {
  cat("Event Study: imputation method. Dep. Var.: ", x$y, "\n")
  cat("Counterfactual model: ", deparse(x$y0), "\n")
  cat("Observations:", nrow(x$data), "\n")
  print(tail(x$coeftable, n = 5L))
  if (nrow(x$coeftable > 5)) {
    cat(" ......", nrow(x$coeftable) - 5, " rows not shown.\n")
  }
}


#' DiD imputation plot
#'
#' @description
#' Takes a didImputation estimation and return a graph of estimated
#' parameters(for dynamic effects only).
#'
#' @param object an object of class DidImputation.
#' @param type Character. Type of plot. Can be either 'default' or 'IR' for
#' impulse response style.
#' @param ci Numeric. Confidence interval( 0.95 by default).
#' @param ... additional arguments.
#'
#' @return A ggplot2 object
#'
#' @import ggplot2
#'
#' @importFrom stats qnorm
#'
#' @examples
#' # standard plot
#' res <- didImputation(y ~ 0 | i + t,
#'               cohort = 'g',
#'               data = did_simulated)
#'
#' didplot(res)
#'
#' # Plot with 99% confidence interval
#'
#' didplot(res, ci = 0.99)
#'
#' # Plot with Impulse response function style
#'
#' didplot(res, type = 'IR')
#'
#'
#' @export
didplot <- function(object,
                    type = "default",
                    ci = 0.95,
                    ...) {
  with(object, {

    # Extract timing (and group in case of triple difference)
    if(object$ncontrast == 1) {
      coeftable$x <- as.numeric(
        sapply(
          strsplit(
            rownames(coeftable), "::"
          ), function(x) x[2]
        )
      )
    } else {
      coeftable$x <- as.numeric(
        sapply(
          strsplit(
            rownames(coeftable), ":"
          ), function(x) x[3]
        )
      )
      coeftable$group <- as.factor(
        sapply(
          strsplit(
            rownames(coeftable), ":"
          ), function(x) x[6]
        )
      )
    }

    q <- qnorm(ci + (1 - ci) / 2)

    ylab <- paste0("Estimate and ", ci * 100, "% Conf. Int.")

    if (!any(coeftable$x == -1, na.rm = TRUE) & min(coeftable$x, na.rm = TRUE) < -1) {
      ref <- data.frame(
        Estimate = 0,
        row.names = paste0("k::", -1),
        x = -1,
        "Std. Error" = NaN,
        "t value" = NaN,
        "Pr(>|t|))" = NaN,
        check.names = FALSE
      )

      coeftable <- rbind(coeftable, ref)
    }


    if(ncontrasts == 1) {
      p <- ggplot(coeftable, aes(x = x, y = Estimate))
    } else {
      p <- ggplot(coeftable, aes(x = x, y = Estimate, color = group))
    }

    p <- p +
      xlab("Time to treatment") +
      ylab(ylab) +
      scale_x_continuous(breaks = seq(min(coeftable$x, na.rm = T), max(coeftable$x, na.rm = T))) +
      theme_classic() +
      theme(
        panel.grid.major.y = element_line(colour = "lightgrey", linetype = "dotted"),
        axis.text = element_text(size = 12)
      )

    if (any(coeftable$x < 0)) p <- p + geom_vline(xintercept = -0.5, color = "firebrick")


    if (type == "default") {
      p <- p +
        geom_hline(yintercept = 0, linetype = "dashed") +
        geom_point() +
        geom_errorbar(aes(
          ymin = Estimate - `Std. Error` * q,
          ymax = Estimate + `Std. Error` * q
        ),
        width = 0.2
        )
    } else if (type == "IR") {
      p <- p +
        geom_hline(yintercept = 0, linetype = "solid") +
        geom_line() +
        if(ncontrasts == 1){
          geom_ribbon(aes(
            ymax = Estimate + `Std. Error` * q,
            ymin = Estimate - `Std. Error` * q,
          ), linetype = 2, alpha = 0, colour = "black")
        } else {
          geom_ribbon(aes(
            ymax = Estimate + `Std. Error` * q,
            ymin = Estimate - `Std. Error` * q,
          ), linetype = 2, alpha = 0)
        }
    }

    return(p)
  })
}

#' DiD imputation plot
#'
#' @inherit didplot
#'
#' @export plot.didImputation
#' @export
#'
plot.didImputation <- function(...) {
  didplot(...)
}
